File: VALIDACION_COTIZADOR_UNIVERSAL.py
# -*- coding: utf-8 -*-
"""Regresiones offline del cotizador universal y sus contratos de interfaz."""
from pathlib import Path
from decimal import Decimal

from cotizacion_document_service import catalogo_companias
from config_service import default_config
from mercantil_cotizador import calcular_descuento
from quote_normalizer import PROFILE_LABELS, PERFIL_LB, PERFIL_LB1


BASE = Path(__file__).resolve().parent
JS = (BASE / "static" / "js" / "app.js").read_text(encoding="utf-8")
HTML = (BASE / "templates" / "documentos.html").read_text(encoding="utf-8")
APP = (BASE / "app.py").read_text(encoding="utf-8")
BASE_HTML = (BASE / "templates" / "base.html").read_text(encoding="utf-8")
ESTILO = (BASE / "static" / "css" / "estilo.css").read_text(encoding="utf-8")
DESIGN = (BASE / "static" / "css" / "design-system.css").read_text(encoding="utf-8")


def check(condicion, mensaje):
    if not condicion:
        raise AssertionError(mensaje)


# ATM nunca muestra ambos catálogos a la vez; desconocido cae conservadoramente
# en auto, por lo que MOTOS tampoco aparece por defecto.
check("matrix.hidden=esMoto" in JS, "ATM auto no oculta la matriz de motos correctamente.")
check("motoRow.hidden=!esMoto" in JS, "ATM moto/auto perdió visibilidad condicional.")
check("==='moto'?'moto':'auto'" in JS, "El tipo ATM desconocido dejó de caer en auto.")

# Mercantil tiene un solo control global y ninguna fila seleccionada crea un
# input por cobertura. La fórmula sigue siendo la misma; 40% debe recalcular.
check(HTML.count('id="mercantilBonificacion"') == 1, "Mercantil no tiene una única bonificación global.")
check("const discount=document.createElement('span')" in JS, "La bonificación volvió a ser editable por fila.")
check("recalcularMercantilTodas" in JS, "La bonificación global no recalcula todas las alternativas.")
for precio in (100000, 179666, 250000):
    resultado = calcular_descuento(precio, 40, 50)
    check(Decimal(resultado["precio_final"]) == Decimal(precio) * Decimal("0.60"), "Mercantil 35→40 no recalculó correctamente.")

# La identidad se edita por source id, no mediante una variable global de
# compañía. Elegir una conocida toma el logo del manifiesto; otra queda textual.
check("fuenteCotizacionPorId" in JS and "confirmarCompaniaFuente" in JS, "Falta identidad de compañía por fuente.")
check("compania_confirmada" in JS and "compania_detectada" in JS, "No se separó compañía detectada de confirmada.")
check("/api/cotizaciones/companias" in APP, "Falta catálogo de logos para Cotizaciones.")
catalogo = catalogo_companias(
    BASE / "static" / "img" / "companias",
    default_config().get("companias"),
)
nombres_catalogo = {item["nombre"] for item in catalogo}
for requerida in ("ATM", "Federación Patronal", "Mercantil Andina", "San Cristóbal", "AgroSalta", "Allianz", "PROF"):
    check(requerida in nombres_catalogo, f"El catálogo universal perdió {requerida}.")
# El logo es un asset opcional: una compañía configurada debe seguir siendo
# seleccionable aunque todavía no tenga logo documental.
for item in catalogo:
    if item.get("file"):
        check((BASE / "static" / "img" / "companias" / item["file"]).is_file(), f"Falta logo declarado de {item['nombre']}.")
check(len([x for x in catalogo if x["nombre"] == "Federación Patronal"]) == 1, "El alias Self duplicó Federación Patronal.")

# UI, WhatsApp y carta parten del mismo objeto comercial de salida.
check("modeloComercialOpcion" in JS, "Falta el adaptador comercial único.")
check("nombre_comercial" in JS and "variante_comercial" in JS and "variante_grua" in JS, "El modelo comercial está incompleto.")
check("modelo.titulo_comercial" in JS, "Las filas seleccionadas no consumen el título comercial.")
check("gruaTokenExplicito" in JS, "Una variante de grúa explícita puede perderse al reconstruir el título.")

check("const resuelto=String(op?.nombre_cliente||'').trim()" in JS, "La familia normalizada vuelve a tener prioridad sobre el nombre resuelto por la fuente.")
check("document.querySelectorAll('.insured-profile-card')" not in JS[JS.find("function posicionarMenuComandos"):JS.find("function renderComandosMenu")], "El menú slash volvió a depender de la posición de la ficha.")
check("setTimeout(posicionarMenuComandos,40)" not in JS, "El menú slash volvió a reposicionarse dinámicamente después del layout.")
check(".chat-command-menu{left:16px;right:auto;bottom:68px" in ESTILO, "El menú slash no quedó anclado al lado izquierdo del composer.")
check("overscroll-behavior:contain" in ESTILO, "El menú slash no contiene correctamente su propio scroll.")

# El menú slash debe leerse de izquierda a derecha: icono, comando y descripción alineados a la izquierda.
CSS_TEXT=(BASE / "static" / "css" / "estilo.css").read_text(encoding="utf-8")
check("justify-content:flex-start!important" in CSS_TEXT, "El menú slash volvió a distribuir icono/texto hacia los extremos.")
check(".chat-command-item strong," in CSS_TEXT and "text-align:left!important" in CSS_TEXT, "Los textos del menú slash no están forzados a la izquierda.")

COT_DOC = (BASE / "cotizacion_document_service.py").read_text(encoding="utf-8")
check('PROGRAMFILES' in COT_DOC and 'LibreOffice" / "program" / "soffice.exe' in COT_DOC, "Cotizaciones no detecta LibreOffice instalado en Windows fuera del PATH.")

# Los perfiles LB/LB1 no pueden volver a exponerse como códigos técnicos.
check(not PROFILE_LABELS[PERFIL_LB].lower().startswith("cobertura lb"), "LB conserva un nombre técnico en vez de comercial.")
check(not PROFILE_LABELS[PERFIL_LB1].lower().startswith("cobertura lb"), "LB1 conserva un nombre técnico en vez de comercial.")
check("LB:'Robo e Incendio + Robo Parcial al Amparo + Accidente Total'" in JS, "Frontend y normalizador no comparten el nombre comercial de LB.")
check("LB1:'Robo e Incendio + Robo Parcial al Amparo'" in JS, "Frontend y normalizador no comparten el nombre comercial de LB1.")

# design-system.css se carga después de estilo.css: la uniformidad definitiva debe
# vivir en la hoja que realmente gana la cascada, no en una regla inefectiva.
check(BASE_HTML.index("css/estilo.css") < BASE_HTML.index("css/design-system.css"), "Cambió el orden esperado de hojas de estilo.")
check("COTIZADOR UNIVERSAL · UNIFORMIDAD FINAL 2026-09-13" in DESIGN, "La capa visual final del cotizador no está en design-system.css.")
check("COTIZADOR UNIVERSAL · UNIFORMIDAD VISUAL / MARCAS 2026-09-13" not in ESTILO, "Quedó una capa duplicada e inefectiva en estilo.css.")
check(".quote-selected-row" in DESIGN and ".quote-source-brand-mark img" in DESIGN, "Faltan reglas comunes para seleccionadas o favicons de compañía.")

# Alta rápida RC: es una entrada manual al MISMO pipeline, no una plantilla
# AgroSalta aislada. Debe permitir compañía + RC / RC sin grúa + precio.
for elemento in ("quotesAddRcBtn", "quotesRcCompany", "quotesRcType", "quotesRcPrice", "quotesRcConfirm"):
    check(elemento in HTML or elemento in JS, f"Falta el control universal de RC: {elemento}")
check("Responsabilidad Civil sin grúa" in HTML, "La variante RC sin grúa no está disponible.")
check("servicio_grua:sinGrua?false:true" in JS, "RC manual no conserva la diferencia real de asistencia/grúa.")
check("notaFranquiciaCotizacion" in JS and "franquicia equivalente al ${porcentaje} de la suma asegurada" in JS, "La franquicia universal no se explica para cualquier porcentaje.")

# /patente /asegurado /info deben figurar como aliases visibles del mismo
# buscador de cartera; ARCA conserva comandos separados.
for comando in ("/patente", "/asegurado", "/info", "/cuit", "/cuil"):
    check(f"comando:'{comando}'" in JS, f"Falta {comando} en el menú de comandos.")

print("OK - cotizador universal: UI común, bonificación global, compañía editable, RC universal, franquicias, aliases y cascada visual")

File: VALIDACION_COTIZADOR_UNIVERSAL.js
/* Regresiones funcionales puras del modelo JS de Cotizaciones, sin navegador. */
const fs = require('fs');
const vm = require('vm');

const noop = () => {};
const storage = {getItem: () => null, setItem: noop, removeItem: noop};
const documentStub = {
  addEventListener: noop,
  querySelector: () => null,
  querySelectorAll: () => [],
  getElementById: () => null,
  documentElement: {getAttribute: () => '', setAttribute: noop, style: {setProperty: noop}},
  body: {classList: {contains: () => false, add: noop, remove: noop, toggle: noop}, appendChild: noop},
};
const context = {
  console,
  document: documentStub,
  localStorage: storage,
  sessionStorage: storage,
  navigator: {},
  location: {pathname: '/', search: '', hash: ''},
  history: {},
  setTimeout,
  clearTimeout,
  URL,
  Blob,
  FormData,
  File: class {},
  window: null,
};
context.window = context;
context.window.addEventListener = noop;
context.window.matchMedia = () => ({matches: false, addEventListener: noop});
context.window.getComputedStyle = () => ({});
vm.createContext(context);
vm.runInContext(fs.readFileSync('static/js/app.js', 'utf8'), context);

function evaluate(source) {
  return vm.runInContext(source, context);
}
function check(condition, message) {
  if (!condition) throw new Error(message);
}

(async () => {
  evaluate(`quoteCompanyCatalog=[
    {key:'agrosalta',nombre:'Agrosalta',logo_url:'/static/img/companias/agrosalta.png',aliases:['AgroSalta','AGROSALTA']},
    {key:'rivadavia',nombre:'Rivadavia',logo_url:'/static/img/companias/rivadavia.png',aliases:['Seguros Rivadavia']},
    {key:'federacion_patronal',nombre:'Federación Patronal',aliases:['Federación','Federacion','Fed','FedPat']},
    {key:'zurich',nombre:'Zurich',aliases:['Zurich Seguros']}
  ]`);
  check(evaluate("resolverCompaniaCotizacion('AGRO SALTA')?.key") === 'agrosalta', 'Falló alias AgroSalta.');
  check(evaluate("resolverCompaniaCotizacion('Seguros Rivadavia')?.logo_url").endsWith('rivadavia.png'), 'Falló logo Rivadavia.');
  const manualFed = evaluate("parsearAlternativaManual('Fed CF SA 71500000 precio 152000')");
  check(manualFed?.compania === 'Federación Patronal' && manualFed?.codigo === 'CF', 'La alternativa manual dejó de reconocer alias central de Federación.');
  const manualZurich = evaluate("parsearAlternativaManual('Zurich Z9 SA 40000000 precio 120000')");
  check(manualZurich?.compania === 'Zurich' && manualZurich?.codigo === 'Z9', 'Una compañía registrada sin parser específico no entra por alternativa manual.');
  check(manualZurich?.descripcion === '', 'Se inventó un speech para una compañía/código sin conocimiento confirmado.');

  evaluate(`quoteSources=[inicializarIdentidadFuente({id:'atm',compania:'ATM',compania_detectada:'ATM',opciones:[
    {codigo:'A',tipo_vehiculo:'auto',nombre_cliente:'Responsabilidad Civil',grua:true},
    {codigo:'A1',tipo_vehiculo:'auto',nombre_cliente:'Responsabilidad Civil',grua:false}
  ]},'ATM')]`);
  check(evaluate("modeloComercialOpcion(quoteSources[0],quoteSources[0].opciones[0]).titulo_comercial") === 'Responsabilidad Civil · CON GRÚA', 'Falta CON GRÚA.');
  check(evaluate("modeloComercialOpcion(quoteSources[0],quoteSources[0].opciones[1]).titulo_comercial") === 'Responsabilidad Civil · SIN GRÚA', 'Falta SIN GRÚA.');
  check(evaluate("modeloComercialOpcion(quoteSources[0],{codigo:'TR',tipo_vehiculo:'auto',nombre_cliente:'Todo Riesgo',franquicia_pct:'3%%'}).titulo_comercial") === 'Todo Riesgo · FRANQUICIA 3%', 'Falló franquicia TR.');

  // La familia normalizada no puede aplastar una variante comercial ya
  // resuelta por la fuente/catálogo (caso ATM Plus/Premium/Black).
  check(evaluate("modeloComercialOpcion(quoteSources[0],{codigo:'CPr',familia:'C_PLUS',perfil_normalizado:'C_PLUS',nombre_cliente:'Terceros Completo Premium'}).titulo_comercial") === 'Terceros Completo Premium', 'ATM Premium volvió a ser aplastado por C_PLUS.');
  check(evaluate("modeloComercialOpcion(quoteSources[0],{codigo:'CB',familia:'C_PLUS',perfil_normalizado:'C_PLUS',nombre_cliente:'Terceros Completo Black'}).titulo_comercial") === 'Terceros Completo Black', 'ATM Black volvió a ser aplastado por C_PLUS.');
  const atmBParcial = evaluate(`datosCoberturaPropuesta({
    compania:'ATM',codigo:'B',familia:'B',nombre:'Robo e Incendio Total y/o Parcial + Accidente Total',
    nombreComercial:'Robo e Incendio Total y/o Parcial + Accidente Total',
    riesgosDetectados:['RESPONSABILIDAD_CIVIL','INCENDIO_TOTAL','INCENDIO_PARCIAL','ROBO_HURTO_TOTAL','ROBO_HURTO_PARCIAL','DESTRUCCION_TOTAL_ACCIDENTE'],
    descripcion:'',precio:'$100.000'
  })`);
  const atmBTextos=atmBParcial.contenidos.filter(x=>x.tipo!=='nota').map(x=>x.texto);
  check(atmBTextos.includes('Incendio Total y Parcial')&&atmBTextos.includes('Robo/Hurto Total y Parcial'), 'La familia B volvió a pisar riesgos estructurados ATM.');

  evaluate(`genericQuoteSources=[inicializarIdentidadFuente({id:'generica-1',compania:'Compañía no identificada',opciones:[]},'Compañía no identificada')];quoteSources.push(genericQuoteSources[0])`);
  evaluate("confirmarCompaniaFuente('generica-1','AgroSalta','agrosalta')");
  check(evaluate("genericQuoteSources[0].compania_confirmada") === 'Agrosalta', 'La confirmada no reemplazó a la detectada.');
  check(evaluate("genericQuoteSources[0].company_key") === 'agrosalta', 'La confirmación no eligió logo/key.');
  evaluate("confirmarCompaniaFuente('generica-1','Rivadavia','rivadavia')");
  check(evaluate("genericQuoteSources[0].company_key") === 'rivadavia', 'El cambio posterior dejó el logo/key anterior.');

  context.fetch = async (_url, options) => {
    const body = JSON.parse(options.body);
    const final = Number(body.precio_base) * (1 - Number(body.descuento) / 100);
    return {ok: true, status: 200, redirected: false, text: async () => JSON.stringify({ok: true, precio_final: String(final)})};
  };
  evaluate(`mercantilFuente={id:'mercantil',compania:'Mercantil Andina',bonificacion:40,opciones:[
    {precio_base:'100000',descuento:35},{precio_base:'200000',descuento:35},{precio_base:'300000',descuento:30}
  ]}`);
  await evaluate('recalcularMercantilTodas()');
  check(evaluate('mercantilFuente.opciones.every(x=>x.descuento===40)') === true, '40% no se aplicó globalmente.');
  check(evaluate("mercantilFuente.opciones.map(x=>x.precio_final).join(',')") === '60000,120000,180000', 'No se recalcularon todas las alternativas.');

  check(evaluate("normalizarPorcentajeCotizacion('2%%')") === '2%', 'Se generó un porcentaje duplicado.');
  const mercantilPlus = evaluate(`datosCoberturaPropuesta({
    compania:'Mercantil Andina',codigo:'MP',familia:'C_PLUS',nombre:'Terceros Completo M Plus',
    nombreComercial:'Terceros Completo M Plus',servicioGrua:true,
    descripcion:'Responsabilidad civil, incendio total y parcial, robo total y parcial, destrucción total por accidente. Cubre ruedas, vidrios, granizo y cerraduras.',
    precio:'$100.000'
  })`);
  check(mercantilPlus.contenidos.some(x => /Incluye grúa/i.test(x.texto)), 'M Plus perdió la grúa confirmada en la salida comercial.');
  const tr75 = evaluate(`datosCoberturaPropuesta({
    compania:'San Cristóbal',familia:'TODO_RIESGO',nombre:'Todo Riesgo',nombreComercial:'Todo Riesgo',
    descripcion:'Responsabilidad civil. Daños parciales por accidente.',franquiciaPct:'7,5',precio:'$200.000'
  })`);
  check(!tr75.contenidos.some(x => x.tipo==='nota' && /franquicia/i.test(x.texto)), 'La nota de franquicia no debe viajar duplicada dentro de contenidos.');
  const wa75 = evaluate(`bloqueCoberturaWhatsAppDatos(${JSON.stringify(tr75)})`);
  check((wa75.match(/franquicia equivalente al 7,5% de la suma asegurada/gi)||[]).length===1, 'WhatsApp debe explicar la franquicia exactamente una vez.');
  const atmTr = evaluate(`datosCoberturaPropuesta({
    compania:'ATM',familia:'TODO_RIESGO',nombre:'Todo Riesgo',nombreComercial:'Todo Riesgo',
    descripcion:'Responsabilidad civil, incendio total y parcial, robo total y parcial, destrucción total y daños parciales por accidente. En caso de daño parcial, queda a cargo del asegurado una franquicia equivalente al 3% de la suma asegurada. Todo gasto que supere ese importe queda a cargo de la compañía.',
    franquiciaPct:'3',mostrarFranquiciaWhatsApp:false
  })`);
  const atmWa = evaluate(`bloqueCoberturaWhatsAppDatos(${JSON.stringify(atmTr)})`);
  check((atmWa.match(/franquicia equivalente al 3% de la suma asegurada/gi)||[]).length===1, 'ATM perdió o duplicó la explicación de franquicia en WhatsApp.');
  check(!/\*Franquicia:\*/i.test(atmWa), 'ATM cambió el formato histórico y agregó una línea de franquicia duplicable.');
  const plusTextos=mercantilPlus.contenidos.filter(x=>x.tipo!=='nota').map(x=>x.texto);
  for(const esperado of ['Responsabilidad Civil','Incendio Total y Parcial','Robo/Hurto Total y Parcial','Destrucción Total por Accidente']){
    check(plusTextos.includes(esperado), `C_PLUS perdió la prestación canónica: ${esperado}.`);
  }
  check(!plusTextos.some(x=>/, incendio/i.test(x)&&/, robo/i.test(x)), 'C_PLUS volvió a una frase legacy en lugar de una prestación por bullet.');

  // ATM: 3% y 6% son alternativas independientes; seleccionar una nunca
  // puede apagar la otra y no existe un selector genérico TR visible.
  evaluate(`atmCoberturasDetectadas=[
    {codigo:'TR',tipo_vehiculo:'auto',franquicia_pct:'3',ofrecer:false,orden:120},
    {codigo:'TR',tipo_vehiculo:'auto',franquicia_pct:'6',ofrecer:false,orden:120}
  ]`);
  evaluate("togglearFranquiciaATM('3')");
  evaluate("togglearFranquiciaATM('6')");
  check(evaluate("atmCoberturasDetectadas.filter(x=>x.codigo==='TR'&&x.ofrecer).length")===2, 'ATM sigue haciendo mutuamente excluyentes las franquicias 3% y 6%.');
  check(evaluate("ATM_CODIGOS_ORDEN.filter(codigo=>codigo!=='TR').includes('TR')")===false, 'El selector genérico TR volvió a la matriz visible.');

  console.log('OK - modelo JS de Cotizaciones, prestaciones canónicas, TR independiente y bonificación global');
})().catch(error => {
  console.error(error);
  process.exit(1);
});
